dozer-sql/src/pipeline/product/join/operator/mod.rs

```diff
@@ -0,0 +1,260 @@
+            accurate_keys,
+        )?;
+        Ok(Self {
+            join_type,
+            left,
+            right,
+        })
+    }
+
+    pub fn left_lookup_size(&self) -> usize {
+        self.left.lookup_size()
+    }
+
+    pub fn right_lookup_size(&self) -> usize {
+        self.right.lookup_size()
+    }
+
+    fn inner_join(
+        &self,
+        action: JoinAction,
+        join_key: &JoinKey,
+        record: &ProcessorRecord,
+        record_branch: JoinBranch,
+        default_if_no_match: bool,
+    ) -> Vec<(JoinAction, ProcessorRecord)> {
+        let table = match record_branch {
+            JoinBranch::Left => &self.right,
+            JoinBranch::Right => &self.left,
+        };
+        let join_records = create_join_records_fn(record, record_branch);
+
+        table
+            .get_matching_records(join_key, default_if_no_match)
+            .map(|matching_record| (action, join_records(matching_record)))
+            .collect()
+    }
+
+    fn outer_join(
+        &self,
+        action: JoinAction,
+        join_key: &JoinKey,
+        record: &ProcessorRecord,
+        record_branch: JoinBranch,
+    ) -> Vec<(JoinAction, ProcessorRecord)> {
+        let (table_to_match, matching_record_branch, table_of_record) = match record_branch {
+            JoinBranch::Left => (&self.right, JoinBranch::Right, &self.left),
+            JoinBranch::Right => (&self.left, JoinBranch::Left, &self.right),
+        };
+        let join_records = create_join_records_fn(record, record_branch);
+
+        // We need to query from the table where this record is from:
+        // - For JoinAction::Insert, did this join key exist before this insert? If not, we need to remove the default record.
+        // - For JoinAction::Delete, does this join key exist after this delete? If not, we need to insert the default record.
+        let need_to_act_on_default_record = match action {
+            JoinAction::Insert => {
+                // Because this record is already inserted, the join key didn't exist before this insert iif the matching count is now 1.
+                table_of_record
+                    .get_matching_records(join_key, false)
+                    .take(2)
+                    .count()
+                    == 1
+            }
+            JoinAction::Delete => {
+                table_of_record
+                    .get_matching_records(join_key, false)
+                    .take(1)
+                    .count()
+                    == 0
+            }
+        };
+
+        let mut output_records = vec![];
+        for matching_record in table_to_match.get_matching_records(join_key, false) {
+            let join_record = join_records(matching_record);
+
+            if need_to_act_on_default_record {
+                let default_join_record = create_join_records_fn(
+                    matching_record,
+                    matching_record_branch,
+                )(table_of_record.default_record());
```

Comment:
Maybe you don't need `matching_record_branch`. It could then be moved outside the loop.
```rust
            let default_join_records = create_join_records_fn(
                table_of_record.default_record(),
                record_branch,
            );
            for ... {
                ...
                if need_to_act_on_default_record {
                    let default_join_record = default_join_records(matching_record);
```

Reply:
Wonderful point.